server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -365,24 +369,83 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
         }
     }
 
+    /**
+     * Merging the provided mappings. Actual merging is done in the raw, non-parsed, form of the mappings. This allows to do a proper bulk
+     * merge, where parsing is done only when all raw mapping settings are already merged.
+     */
+    @SuppressWarnings("unchecked")
+    public DocumentMapper merge(String type, List<CompressedXContent> mappingSources, MergeReason reason) {
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null && mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+            return currentMapper;
+        }
+
+        Map<String, Object> mergedRawMapping = null;
+        for (CompressedXContent mappingSource : mappingSources) {
+            Map<String, Object> rawMapping = MappingParser.convertToMap(mappingSource);
+            Iterator<String> iterator = rawMapping.keySet().iterator();
+            if (iterator.hasNext()) {
+                String rootName = iterator.next();
+                if (mappingParser.isRootMappingType(rootName, type)) {
+                    rawMapping = (Map<String, Object>) rawMapping.get(rootName);
+                }
```

Comment:
Maybe the opposite is the right thing to do 🤔 
This may result with a multi-root mapping, so maybe what we should really do is make sure all raw merges start with a common root (eg `_doc`), in other words- add `_doc` parent if it is not the root, rather than remove it if is the root

Reply:
Yep, the opposite is better- I eventually decided to normalize the mappings so that all have the provided `type` argument as their single common root.